### PR TITLE
[SPARK-45856] Move ArtifactManager from Spark Connect into SparkSession (sql/core)

### PR DESCRIPTION
**Pull request by vicennial**
### What changes were proposed in this pull request?

The significant changes in this PR include:

- `SparkConnectArtifactManager` is renamed to `ArtifactManager` and moved out of Spark Connect and into `sql/core` (available in `SparkSession` through `SessionState`) along with all corresponding tests and confs.
 - While `ArtifactManager` in part of SparkSession, we keep the legacy behaviour for non-connect spark while utilising the ArtifactManager in connect pathways
    - This is done by exposing a new method `withResources` in the artifact manager that sets the context class loader (for driver-side operations) and propagates the `JobArtifactState` such that the resources reach the executor.
    - Spark Connect pathways utilise this method through the `SessionHolder#withActive`
    - When `withResources` is not used, neither the custom context classloader nor the `JobArtifactState` is propagated and hence, non Spark Connect pathways remain with legacy behaviour.  

### Why are the changes needed?
The `ArtifactManager` that currently lies in the connect package can be moved into the wider sql/core package (e.g SparkSession) to expand the scope. This is possible because the `ArtifactManager` is tied solely to the `SparkSession#sessionUUID` and hence can be cleanly detached from Spark Connect and be made generally available.

### Does this PR introduce _any_ user-facing change?

No. Existing behaviour is kept intact for both non-connect and connect spark.

### How was this patch tested?

Existing tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by vicennial** [comment]
PTAL @hvanhovell @HyukjinKwon

**Review by hvanhovell — APPROVED** [review]
LGTM

**Review by dongjoon-hyun — COMMENTED** [review]
According to the JIRA, this is only for Apache Spark 4.0.0, right?

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 1]
```diff
@@ -243,6 +244,16 @@ class SparkSession private(
   @Unstable
   def streams: StreamingQueryManager = sessionState.streamingQueryManager
 
+  /**
+   * Returns an `ArtifactManager` that supports adding, managing and using session-scoped artifacts
+   * (jars, classfiles, etc).
+   *
+   * @since 3.5.1
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:251]: This should be 4.0.0 because this PR is for Apache Spark 4.0.0, @vicennial .

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
We still need to support configurations in a way because we didn't deprecate yet in the Apache Spark community.
```
spark.connect.copyFromLocalToFs.allowDestLocal
```

**Comment by dongjoon-hyun** [comment]
Could you re-trigger the failed pipelines?

**Comment by vicennial** [comment]
@dongjoon-hyun Thank you for the review! 

I've updated the version and for the deprecated conf `spark.connect.copyFromLocalToFs.allowDestLocal`, I've added it to the `deprecatedSQLConfigs` list so that it would generate a warning if used. Currently, both configs would work but if the new config is set (currently optional), it would override the deprecated config. Eventually, we should remove the deprecated conf and convert the new config from optional to have a default value.

**Comment by vicennial** [comment]
Hmm, JavaDocGeneration is failing in 2 tests and from the logs, its not clear why... 
I will merge master and see if that helps

**Comment by vicennial** [comment]
@dongjoon-hyun The CI is green now :)

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by fhalde** [comment]
Hi, we are super interested in having the isolated classloader per spark session ability for our usecase. i believe this today is only achievable if jobs are run from a connect client. we want to avoid using connect client

but with this pr merged, it should be possible to have isolated classloaders per spark session right without any use of spark connect?

our use-case involves starting a spark driver and dynamically loading & adding jars and running transformations present within the jars. without isolated classloaders per session, on the executor side we would risk classpath conflict

@HyukjinKwon if your pr addresses my concern, i can back port it to 3.5.0 in my fork

**Comment by vicennial** [comment]
@fhalde Yes, with this PR, it would be possible to have isolated classloaders per spark session on the executors without going through Spark Connect. 

The `withResources` method should be used to wrap all executions (like in Spark Connect [here](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala#L245)) and note that, all artifacts (i.e Jars, classfiles) would need to be added through the `ArtifactManager` (directly adding these to `SparkContext` would not work), refer to Spark Connect's `AddArtifactsHandler` to see how we use the [API](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala#L90).

File: project/MimaExcludes.scala [hunk 2]
```diff
@@ -90,6 +90,16 @@ object MimaExcludes {
     // SPARK-43299: Convert StreamingQueryException in Scala Client
     ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.streaming.StreamingQueryException"),
 
+    // SPARK-45856: Move ArtifactManager from Spark Connect into SparkSession (sql/core)
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.storage.CacheId.apply"),
```
Comment (LuciferYang) [inline project/MimaExcludes.scala:94]: @vicennial I would like to reconfirm, the `ProblemFilters` added by SPARK-45856 will never need to undergo a mima check in versions after Spark 4.0, is that correct? I found that it has been placed in `defaultExcludes`.

**Pull request closed without merging**